### ISSUE TITLE
Success Criterion 2.4.3 - Focus Order - Level A

### ISSUE THREAD

**Issue by JJdeGroot** [w3c/matf#35]
WCAG2ICT guidance: https://www.w3.org/TR/wcag2ict-22/#focus-order

Share your thoughts for applying to mobile apps as a comment below.

**Comment by Keanem6** [comment]
I think it applies as written in terms of requirements. But i do think it's worth what's expected from the basic keystrokes for iOS and Android, and certain patterns, such as where focus should land when navigating to new views/screen, selecting back buttons, closing modals...etc.

**Comment by arctouch-gleidsonramos** [comment]
Remembering that the recommendation for both iOS and Android is that the focus order not be changed programmatically, and that this must be planned by design.

**Comment by jamieherrera** [comment]
This sc is in the top 5 most filed at my company (in both web and mobile) but also most cancelled; "meaningful" could have a note (also apply to 1.3.2 meaningful sequence) to clarify what makes this SC not met, where "meaningful" is maybe more broadly applied in native mobile. 
I'm not as well-versed in the functional ability in native code to alter the tab order; I do know that custom components need to be made keyboard accessible otherwise they may be skipped entirely. I don't think this SC needs to bend to current Apple or Android limitations, but rather include that a default focus order is not a fail.

**Comment by JJdeGroot** [comment]
- [ACTION 1](https://www.w3.org/2024/08/21-matf-minutes.html#a01): Link to AGWG consensus for 2.4.3 Focus Order / 1.3.2 Meaningful Sequence (@detlevhfischer)
- [ACTION 2](https://www.w3.org/2024/08/21-matf-minutes.html#a04): Define what "default" means in context of focus order

**Comment by JJdeGroot** [comment]
Discussed in today’s meeting.

<details>
<summary>15 January 2025</summary>

[Source](https://www.w3.org/2025/01/15-matf-minutes.html#206c)

~~~text
Aash shared that the default flow should be derived from research. Can be a note

Detlev we can't assume that we can reach all elements by tapping / swiping

@detlev if you can reach elements by any means, the issues is pass, but with a note

JJ we need to understand about the nuances of iOS and Android

Jamie we need to maybe differentiate the interactions in keyboard vs non-keyboard

JJ we had an idea to update keyboard interface definition to include all assistive tech

JJ mentioned that keyboard + AT need to be mapped properly as they might interfere with each other

ACTION: Update "navigated sequentially" definition, it mentions "keyboard interface"

ACTION: Define the ways of interaction, e.g. tab, shift+tab, arrow keys, etc. that are acceptable to navigate

Detlev we can define one way where all elements can be reached with the same interaction, arrow or tab

ACTION: Think about what "sequential" means in a mobile context in relation with the keys used for navigating

Detlev the special keystrokes needed to go to navigation / header might not be known to all.

Jamie Agree with Detlev . Special keystrokes may confuse people.

Jamie the tester needs to know how the keyboard interacts with the operating system

ACTION: Consider adding Note that failure depends on whether the expected keyboard operation matches the actual keyboard operation of the component / screen

<julianmka> +1 Jamie

Jamie we may add a note about keyboard functionality . That will help reduce confusion

<JJ> julianmka: Take semantics of the view into account to define the expected keyboard operation keys

<julianmka> Workday's research on keyboard nav in mobile: https://github.com/workday-accessibility/accessibility-eval/blob/main/keyboard.md

<Detlev> julianmka I was not suggesting that expectations should be overwritten, sure

<julianmka> Apologies for the misunderstanding Detlev

we can mention the accepted ways of using keyboard based on comments from julianmka and Jamie

JJ the guidelines with medium / large variation, we may have to work more to clarify
~~~
</details>

Actions:
- ACTION: Update "navigated sequentially" definition, it mentions "keyboard interface"
- ACTION: Define the ways of interaction, e.g. tab, shift+tab, arrow keys, etc. that are acceptable to navigate
- ACTION: Think about what "sequential" means in a mobile context in relation with the keys used for navigating
- ACTION: Consider adding Note that failure depends on whether the expected keyboard operation matches the actual keyboard operation of the component / screen

**Comment by JJdeGroot** [comment]
Discussed in today’s meeting.

<details>
<summary>30 April 2025</summary>

[Source](https://www.w3.org/2025/04/30-matf-minutes.html#f3eb)

~~~text
Aash From a design point of view - what if there are elements that cannot be used by (e.g. visual impairments) - for example maps

Sorry I couldn't capture the end of that - someone enabled translations on zoom and zoom stole my focus

Which is ironic

<JJ> sorry, just opened the transcript panel to keep track of all text

Aash maps iterate through focus which is not helpful for people with assistive technologies.

pauljadam everything operable with a touch gesture needs to be focusable as well

<pauljadam> Put the map in a container with a name and VoiceOver can skip over it or to it

Detlev If I understand correctly - this is not specific to mobile. Maps can have 100's of dots, which can also be problematic due to grouping, it really depends what you do outside that map. I think it's difficult to tell without entire context

<pauljadam> On desktop you can make the map pins focusable with arrow keys like radio buttons rather than tab key so they can skip past faster with keyboard.

<pauljadam> its often confused with reading order yes

Detlev Do we have concensus that focus order does refer to keyboard as opposed to swipe focus order (screen reader)

<JJ> https://www.w3.org/TR/WCAG22/#dfn-navigated-sequentially

@JJ it seems to apply to all

pauljadam I have a focus order test case: as a keyboard user that focus is moved to controls. For screen reader and switch you focus on everything, but keyboard you just get interactive elements. Other assistive tech can be used to test

Important note: Switch ordering and keyboard ordering is different on Android

Switch allows for focus of containers and has an additional menu

ACTION: If keyboard interface is replaced with accessibility interface, emphasize the focus logic / differences between keyboard/web and app behavior

(and actions, don't forget the actions - hidden from keyboard users)

Detlev to what extent do we take modifiers - they are quite common. Is there is a good way of securing focus order without modifiers?

TAB+X is the accessibility shortcut

I really dislike that

(or Z, but it's Tab as a modifier)
~~~
</details>

- ACTION: If keyboard interface is replaced with accessibility interface, emphasize the focus logic / differences between keyboard/web and app behavior

**Comment by mattyzenny** [comment]
I'm curious if Floating Action Buttons (FAB) have been discussed in depth due to their prevalence in Mobile apps, and how they should be handled. 

For instance, how practical would it be to have the FAB at the end of the DOM of an infinite scrolling screen?

**Comment by JJdeGroot** [comment]
> I'm curious if Floating Action Buttons (FAB) have been discussed in depth due to their prevalence in Mobile apps, and how they should be handled.
> 
> For instance, how practical would it be to have the FAB at the end of the DOM of an infinite scrolling screen?

Hi @mattyzenny I think we have discussed floating action buttons in the past. This success criterion hasn't been on the agenda for a year or so. It will be back on the agenda in the second half of this year, when we start the third milestone.

**Comment by mattyzenny** [comment]
> > I'm curious if Floating Action Buttons (FAB) have been discussed in depth due to their prevalence in Mobile apps, and how they should be handled.
> > For instance, how practical would it be to have the FAB at the end of the DOM of an infinite scrolling screen?
> 
> Hi [@mattyzenny](https://github.com/mattyzenny) I think we have discussed floating action buttons in the past. This success criterion hasn't been on the agenda for a year or so. It will be back on the agenda in the second half of this year, when we start the third milestone.

Thanks for the response. I look forward to hearing the results of the future discussions.
